Generate Makefiles

njuettner opened the pull request:
Add to generate Makefiles for Golang projects.

This is a followup from the conversation with Pawel about adding a Makefile for the aws-operator (https://github.com/giantswarm/aws-operator/pull/2445). I hope this PR is a better approach for using Makefiles 😄.

But again this PR is open for discussions.

kopiczko commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+SOURCES       = $(shell find . -name '*.go')
+GOPKGS        = $(shell go list ./...)
+TAG           ?= $(VERSION)
+BUILD_FLAGS   ?= -v
```
I'd skip the `-v` flag. It's noisy.

kopiczko replied:
It would be good to make it on par with the [orb](https://github.com/giantswarm/architect-orb/blob/7e3e1243d49a67d6f07482e40145e3da3598ded2/src/commands/go-test.yaml#L5-L9). But it's fine with me if we don't do this in first iteration.

kopiczko commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+	Application string
+}
+
+var Makefile = `APPLICATION   ?= {{.Application}}
```
```suggestion
var Makefile = `# DO NOT EDIT. Generated with:
#
#    devctl gen makefile
#

APPLICATION   ?= {{.Application}}
```

kopiczko commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+type TemplateConfig struct {
+	Application string
+}
+
```
Because of consistency I would be happier if we followed the pattern like [here](https://github.com/giantswarm/devctl/blob/master/pkg/gen/input/ami/internal/file/ami.go).

Even though I understand it may feel like an overkill for this particular case.

If you'd like to do so then we need to make this file regenerable [here](https://github.com/giantswarm/devctl/blob/c5630628e98c5fb33582f1511a8f02332c90b50d/pkg/gen/gen.go#L44-L52).

But I won't block on that because I think it's already an improvement.

kopiczko commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+var Makefile = `APPLICATION   ?= {{.Application}}
+VERSION       ?= $(shell git describe --tags --always --dirty)
```
```suggestion
VERSION       ?= $(shell architect project version)
```

kopiczko commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+.PHONY: test
+## test: runs go test with default values
+test:
+	go test -v -race -cover $(GOPKGS)
```
I'd drop `-v` here as well. We've been in a place where we had so many packages that it was difficult to find what was failing when tests were red.

kopiczko commented on cmd/gen/makefile/flag.go:
```diff
@@ -0,0 +1,29 @@
+
+func (f *flag) Init(cmd *cobra.Command) {
+	cmd.Flags().StringVar(&f.App, app, "", `Name of the application, this will be the name of the go binary.`)
+	cmd.Flags().StringVar(&f.Output, "output", "", `Where you want the generated Makefile to be saved`)
```
That flat I'd skip unless there is a use-case for that. We should generate it as `Makefile` in the current dir.

kopiczko commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+
+var Makefile = `APPLICATION   ?= {{.Application}}
+VERSION       ?= $(shell git describe --tags --always --dirty)
+SOURCES       = $(shell find . -name '*.go')
```
Do we need sources when we have only phony targets?

fiunchinho commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+VERSION       ?= $(shell git describe --tags --always --dirty)
+SOURCES       = $(shell find . -name '*.go')
+GOPKGS        = $(shell go list ./...)
+TAG           ?= $(VERSION)
```
Why have both `VERSION` and `TAG`?

fiunchinho commented on cmd/gen/makefile/template/template.go:
```diff
@@ -0,0 +1,67 @@
+package template
```
What do you folks think about running `go mod tidy` automagically after building?

fiunchinho commented:
This looks great! It would be cool to try to replicate [our orb](https://github.com/giantswarm/architect-orb/blob/a4a74fed9a3a719ed7e62b0c16bcb866a08838d1/src/commands/go-test.yaml) as much as possible, so that we can later use the Makefile from the orb instead.

njuettner commented on cmd/gen/makefile/runner.go:
```diff
@@ -0,0 +1,64 @@
+		return microerror.Mask(err)
+	}
+
+	err = gen.Execute(
```
@kopiczko I moved the logic to your suggested place. How can I avoid getting the `zz_generated.Makefile` file and get a `Makefile` file? TBH I underestimated the complexity of this logic and I'm not sure how to make this right 🙂 . I would appreciate your input.

kopiczko replied:
How about that https://github.com/giantswarm/devctl/pull/57 ?

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,84 @@
+#    devctl gen makefile
+#
+
+APPLICATION    ?= {{.Application}}
```
```suggestion
APPLICATION    ?= $(shell basename $(go list .))
```

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,84 @@
+BUILDTIMESTAMP = $(shell date -u '+%FT%TZ')
+LDFLAGS        ?= -w -linkmode 'auto' -extldflags '-static' \
+  -X '$(go list .)/pkg/project.buildTimestamp=${BUILDTIMESTAMP}' \
+  -X '$(go list .)/pkg/project.gitSHA=${GITSHA1}'
```
```suggestion
  -X '$(go list .)/pkg/project.gitSHA=${GITSHA1}' \
  -X '$(go list .)/pkg/project.version=${VERSION}'
```

njuettner commented:
I cleaned it up a bit and dropped the flag since we don't need it anymore.

Had to change `$(shell basename $(go list .))`to  `$(shell basename $(shell go list .))` other wise it was complaining on my Mac that basename not getting the string.

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,82 @@
+  -X '$(go list .)/pkg/project.buildTimestamp=${BUILDTIMESTAMP}' \
+  -X '$(go list .)/pkg/project.gitSHA=${GITSHA1}' \
+  -X '$(go list .)/pkg/project.version=${VERSION}'
+default: build
+
+.PHONY: build
+## build: builds a local binary
+build: clean
+	CGO_ENABLED=0 go build -o ${APPLICATION} .
+
+.PHONY: build-linux
+## build-linux: builds binary for linux/amd64
+build-linux: clean
+	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 go build .
+
+.PHONY: build-darwin
+## build-darwin: builds binary for darwin/amd64
+build-darwin: clean
+	GOOS=darwin GOARCH=amd64 CGO_ENABLED=0 go build .
```
```suggestion
.DEFAULT_GOAL := build

.PHONY: build build-darwin build-linux
## build: builds a local binary
build: $(APPLICATION)
## build-darwin: builds a local binary for darwin/amd64
build-darwin: $(APPLICATION)-darwin
## build-linux: builds a local binary for linux/amd64
build-linux: $(APPLICATION)-linux

$(APPLICATION): $(APPLICATION)-$(OS)
	cp -a $< $@

$(APPLICATION)-%: $(SOURCES)
	GOOS=$* GOARCH=amd64 go build -ldflags "$(LDFLAGS)" -o $@ .
```

njuettner replied:
@kopiczko I tried that on my mac, `make build-linux` and `make build-darwin` just works but it doesn't work with `make build` or `make` directly.

```
cmd/go: unsupported GOOS/GOARCH pair .o/amd64
make: *** [devctl-.o] Error 2
```

kopiczko replied:
Can you paste the full outpu?

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,82 @@
+.PHONY: install
+## install: install the application
+install:
+	go install .
```
```suggestion
	go install -ldflags "$(LDFLAGS)" .
```

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,82 @@
+.PHONY: run
+## run: runs go run main.go
+run:
+	go run -race .
```
```suggestion
	go run -ldflags "$(LDFLAGS)" -race .
```

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,82 @@
+.PHONY: clean
+## clean: cleans the binary
+clean:
+	go clean
```
```suggestion
	rm -f $(APPLICATION)*
	go clean
```

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,82 @@
+.PHONY: test
+## test: runs go test with default values
+test:
+	go test -ldflags $(LDFLAGS) -race -cover ./...
```
```suggestion
	go test -ldflags "$(LDFLAGS)" -race -cover ./...
```

kopiczko commented on pkg/gen/input/makefile/internal/file/makefile.go:
```diff
@@ -0,0 +1,84 @@
+.PHONY: test
+## test: runs go test with default values
+test:
+	go test -ldflags "$(LDFLAGS)" -race -cover ./...
```
Do we need `-cover` flag there? We don't use it in the CI and I'm not sure it's useful as we are not super fussed about test coverage in general. We also don't have `-race` but I think that's ok if it doesn't take ages to run it in the CI machine.

https://github.com/giantswarm/architect-orb/blob/d1535b34fad4df844df9f183325fde5c24b28ba8/src/commands/go-test.yaml#L23

njuettner replied:
@kopiczko I dropped it but left -race. I think this shouldn't increase the amount of time when running it.

kopiczko replied:
Fine with me.
